src/com/ms/silverking/cloud/dht/daemon/storage/StorageModule.java

```diff
@@ -150,425 +200,545 @@ public StorageModule(NodeRingMaster2 ringMaster, String dhtName, Timer timer, Zo
+            nsLinkBasePath, this, reapPolicy, nsProperties);
+        namespaces.put(ns, nsStore);
+        addNSMetricsNamespaceForNS(nsStore);
+        nsStore.startWatches(zk, nsLinkBasePath, this);
+        Log.warning("\t\tDone recovering: " + nsDir.getName());
+      }
+    } catch (NumberFormatException nfe) {
+      nfe.printStackTrace();
+      Log.warning("Recovery ignoring unexpected nsDir: ", nsDir);
+    }
+  }
+
+  private List<File> sortNSDirsForRecovery(File[] files) throws IOException {
+    List<File> sorted;
+
+    sorted = new ArrayList<>();
+    for (File file : files) {
+      if (file.isDirectory() && !file.getName().equals(trashManualDirName)) {
+        addDirAndParents(sorted, file);
+      } else {
+        Log.warning("Recovery ignoring: ", file);
+      }
+    }
+    return sorted;
+  }
+
+  private void addDirAndParents(List<File> sorted, File childDir) throws IOException {
+    if (!sorted.contains(childDir)) {
+      NamespaceProperties nsProperties;
+      String parentName;
+
+      nsProperties = nsMetaStore.getNsPropertiesForRecovery(childDir);
+      parentName = nsProperties.getParent();
+      if (parentName != null) {
+        long parentContext;
+        File parentDir;
+
+        parentContext = new SimpleNamespaceCreator().createNamespace(parentName).contextAsLong();
+        parentDir = new File(baseDir, NamespaceUtil.contextToDirName(parentContext));
+        addDirAndParents(sorted, parentDir);
+      }
+      sorted.add(childDir);
     }
-    
-    public NamespaceProperties getNamespaceProperties(long ns, NamespaceOptionsRetrievalMode retrievalMode) {
-        return nsMetaStore.getNamespaceProperties(ns, retrievalMode);
+  }
+
+  @Override
+  public ManagedNamespaceStore getManagedNamespaceStore(String nsName) throws ManagedNamespaceNotCreatedException {
+    try {
+      return getNamespaceStore(NamespaceUtil.nameToContext(nsName), NSCreationMode.CreateIfAbsent);
+    } catch (NamespaceNotCreatedException nnce) {
+      throw new ManagedNamespaceNotCreatedException(nnce);
+    }
+  }
+
+  public void ensureMetaNamespaceStoreExists() {
+    metaNamespaceStore = namespaces.get(NamespaceUtil.metaInfoNamespace.contextAsLong());
+    if (metaNamespaceStore == null) {
+      metaNamespaceStore = getNamespaceStore(NamespaceUtil.metaInfoNamespace.contextAsLong(),
+          NSCreationMode.CreateIfAbsent);
     }
-    
-    private NamespaceStore getNamespaceStore(long ns, NSCreationMode mode) {
-        NamespaceStore  nsStore;
-        
+  }
+
+  public NamespaceProperties getNamespaceProperties(long ns, NamespaceOptionsRetrievalMode retrievalMode) {
+    return nsMetaStore.getNamespaceProperties(ns, retrievalMode);
+  }
+
+  private NamespaceStore getNamespaceStore(long ns, NSCreationMode mode) {
+    NamespaceStore nsStore;
+    File nsDir;
+
+    nsDir = new File(baseDir, NamespaceUtil.contextToDirName(ns));
+    nsStore = namespaces.get(ns);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `get()` and `get()` may not be thread-safe at lines: **495, 496, and 508**. In between the check and the `get()` another thread can remove the key and the `get()` will return `null`. The remove that can remove the key is at line: **290**.

**Fix**  
 Consider calling `get()`, checking instead of your current check if the returned object is `null`, and then using that object only, without calling `get()` again.

**More info**  
 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/S3Guard.java#L302-L304) (external link).